src/org/rascalmpl/interpreter/utils/JavaCompiler.java

```diff
@@ -561,6 +596,51 @@ final class ClassLoaderImpl extends ClassLoader {
       super(parentClassLoader);
    }
 
+   public void outputClassesToJar(String qualifiedClassName, OutputStream output) throws IOException {
+      Manifest manifest = new Manifest();
+      manifest.getMainAttributes().put(Attributes.Name.MANIFEST_VERSION, "1.0");
+      manifest.getMainAttributes().put(Attributes.Name.MAIN_CLASS, qualifiedClassName);
+   
+      try (JarOutputStream target = new JarOutputStream(output, manifest)) {
+         for (Entry<String, JavaFileObject> entry : classes.entrySet()) {
+            String className = entry.getKey();
+            JavaFileObjectImpl file = (JavaFileObjectImpl) entry.getValue();
+            JarEntry jarEntry = new JarEntry(className);
+            jarEntry.setTime(file.getLastModified());
+            target.putNextEntry(jarEntry);
+            target.write(file.getByteCode());
+            target.closeEntry();
+         }
+      }  
+   }
+
+   public Class<?> inputClassesFromJar(InputStream in) throws IOException, ClassNotFoundException {
+      try (JarInputStream jarIn = new JarInputStream(in)) {
+         Manifest mf = jarIn.getManifest();
+         String mainClass = (String) mf.getMainAttributes().get(Attributes.Name.MAIN_CLASS);
+         JarEntry jarEntry;
+
+         if (mainClass == null) {
+            throw new IOException("missing Main-Class in jar manifest");
+         }
+
+         while ((jarEntry = jarIn.getNextJarEntry()) != null) {
+            if (!jarEntry.isDirectory()) {
+               var className = jarEntry.getName();
+               var file = new JavaFileObjectImpl(className, JavaFileObject.Kind.CLASS);
+               
+               try (var fo = file.openOutputStream()) {
+                  fo.write(Prelude.consumeInputStream(jarIn));
```

Comment:
we could use a regular copy loop, to avoid loading it all into memory first?

Reply:
yes, but so far this code has proven to be blazingly fast so I thought this is nice and short as it is. We average around 4 ms for storing or reading a medium sized grammar (so that includes reading and writing all byte arrays for the JVM bytecode of all sub-classes and the main parser class.